### PR TITLE
[WIP] Ice block

### PR DESCRIPTION
**Pull request by Pedrokase**
Creates a wall of ice, replacing AIR blocks in given region. An //undo is done after duration ends or any outside source attempts to break the Ice blocks (explosion, piston, player break)

### CODE REVIEW

File: src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java [hunk 1]
```diff
@@ -14,13 +13,15 @@
...
-import utils.selectors.creator.BlockCreator;
+import utils.worldEdit.creators.BlockCreationRequester;
+import utils.worldEdit.creators.TemporaryBlockCreator;
+import utils.worldEdit.selectors.Cuboid2DRegionSelector;
 
 import java.util.Map;
 
 public class IceBlock extends Skill {
     private static final Logger LOGGER = LogManager.getLogger(IceBlock.class);
+
```
Comment (Phobossus) [inline src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java]: test this line

File: src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java [hunk 2]
```diff
@@ -68,13 +69,7 @@ protected void invoke(WorldEntity caster) {
...
 
-        try {
-            BlockCreator blockCreator = new BlockCreator(region);
-            blockCreator.setBlocks(Material.PACKED_ICE);
-            blockCreator.makePermanent();
-
-        } catch (WorldEditException e) {
-            LOGGER.warn("Failed to set Ice blocks using WorldEdit when casting {}, reason: {}", NAME, e.getMessage());
-        }
+        TemporaryBlockCreator blockCreator = new TemporaryBlockCreator(region.getWorld(), 10);
```
Comment (Phobossus) [inline src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java]: extract the magic number

File: src/main/java/utils/worldEdit/TemporaryBlockController.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+public class TemporaryBlockController {
+    //TODO Remove from set once tasks are done
+    private final Set<TemporaryBlockCreator> activeTemporaryCreations = new HashSet<>();
+
+    public void process(BlockCreationRequestEvent event) {
+        BlockCreator blockCreator = event.getBlockCreator();
+        if (event.isCancelled() || !(blockCreator instanceof TemporaryBlockCreator)) {
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/TemporaryBlockController.java]: extract the full condition into a method decribing what its checking (i.e isTemporaryBlock())

Comment (Phobossus) [reply src/main/java/utils/worldEdit/TemporaryBlockController.java]: or isTemporaryBlockCreation(), whatever fits best

File: src/main/java/utils/worldEdit/TemporaryBlockController.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            return;
+        }
+        activeTemporaryCreations.add((TemporaryBlockCreator) blockCreator);
+    }
+
+    public void process(BlockBreakEvent event) {
+        BlockVector3 blockVector = BukkitAdapter.asBlockVector(event.getBlock().getLocation());
+        activeTemporaryCreations.forEach(creation -> {
+            boolean rollbackSuccessful = creation.undoSpecificLocation(blockVector);
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/TemporaryBlockController.java]: undoSpecificBlock?

File: src/main/java/utils/worldEdit/creators/BlockCreator.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+import org.apache.logging.log4j.Logger;
+import org.bukkit.Material;
+import utils.worldEdit.model.ChangeList;
+
+import java.security.InvalidParameterException;
+import java.util.LinkedList;
+
+/**
+ * Use {@link BlockCreationRequester} to request block changes instead, so that events are triggered
+ */
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/creators/BlockCreator.java]: Does the mix of package access and public methods make sense? Perhaps if BlockCreator is now supposed to only be used indirectly, all methods could have package access or "undo" methods should be moved elswhere?

File: src/main/java/utils/worldEdit/events/BlockRemovalEvent.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package utils.worldEdit.events;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class BlockRemovalEvent extends Event {
+    @Override
+    public HandlerList getHandlers() {
+        return null;
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/events/BlockRemovalEvent.java]: D E R P

File: src/main/java/utils/worldEdit/model/BlockCreationRequestEventListener.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    private final TemporaryBlockController temporaryBlockController;
+
+    public BlockCreationRequestEventListener(TemporaryBlockController temporaryBlockController) {
+        this.temporaryBlockController = temporaryBlockController;
+    }
+
+    @EventHandler
+    public void onBlockCreation(BlockCreationRequestEvent event) {
+        LOGGER.info("New block creation request done by '{}", event.getWhoRequested());
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/model/BlockCreationRequestEventListener.java]: Change to "debug" later on.

File: src/main/java/utils/worldEdit/model/ChangeList.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import com.sk89q.worldedit.history.change.Change;
+import com.sk89q.worldedit.history.changeset.ChangeSet;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Implements {@link ChangeSet} and not {@link com.sk89q.worldedit.history.changeset.ArrayListHistory}
+ * because the List field is private and we need it for
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/model/ChangeList.java]: For?

File: src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java (hunk 2 above)
Comment (Pedrokase) [reply src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java]: It's under DURATION_SECONDS, I just used '10' for testing purposes for now

File: src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                .name(NAME)
+                .tier(Tier.I)
+                .description(DESCRIPTION)
+                .addTrigger(TriggerType.PSIONIC, new Trigger("icio"))
+                .build();
+    }
+
+    @Override
+    protected void invoke(WorldEntity caster) {
+//        LivingCharacter livingCharacter = (LivingCharacter) caster;
```
Comment (Phobossus) [inline src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java]: If this code is unnecessary, remove it

File: src/main/java/utils/worldEdit/selectors/Cuboid2DRegionSelector.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+     * Creates a 2D wall that is expanded by 'width' and 'height' from center location.
+     * The walls direction will change depending on center pitch value
+     * <p>
+     * Example (width = 1, height = 1), where O is center:
+     * XXX
+     * XOX
+     * <p>
+     * A value of '0' will make the wall not grow in that axis.
+     */
+    public Region selectWallRegionBasedOnCenter(Location center, FacingLocation facingLocation, int width, int height) {
```
Comment (Phobossus) [inline src/main/java/utils/worldEdit/selectors/Cuboid2DRegionSelector.java]: select2dWallRegion

Comment (Pedrokase) [reply src/main/java/utils/worldEdit/selectors/Cuboid2DRegionSelector.java]: **select2dWallRegionBasedOnCenter** or just **select2dWallRegion**?

Comment (Phobossus) [reply src/main/java/utils/worldEdit/selectors/Cuboid2DRegionSelector.java]: Just that, if we're passing Location center then it's obvious we base it on center

Comment (Pedrokase) [reply src/main/java/utils/worldEdit/selectors/Cuboid2DRegionSelector.java]: Done

File: src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java (hunk 9 above)
Comment (Pedrokase) [reply src/main/java/abilitiesV2/skills/abitilies/manipulation/IceBlock.java]: Done

File: src/main/java/utils/worldEdit/model/ChangeList.java (hunk 8 above)
Comment (Pedrokase) [reply src/main/java/utils/worldEdit/model/ChangeList.java]: Updated, has different name

File: src/main/java/utils/worldEdit/model/BlockCreationRequestEventListener.java (hunk 7 above)
Comment (Pedrokase) [reply src/main/java/utils/worldEdit/model/BlockCreationRequestEventListener.java]: Done

**Pull request merged**